lib/fpm/package/snap.rb

```diff
@@ -0,0 +1,130 @@
+require "fileutils"
+require "fpm/package/dir"
+
+# Support for snaps (.snap files).
+#
+# This supports the input and output of snaps.
+class FPM::Package::Snap < FPM::Package
+
+  option "--yaml", "FILEPATH",
+    "Custom version of the snap.yaml file." do | snap_yaml |
+    File.expand_path(snap_yaml)
+  end
+
+  option "--confinement", "CONFINEMENT",
+    "Type of confinement to use for this snap.",
+    default: "devmode" do | confinement |
+    if ['strict', 'devmode', 'classic'].include? confinement
+      confinement
+    else
+      raise "Unsupported confinement type '#{confinement}'"
+    end
+  end
+
+  option "--grade", "GRADE", "Grade of this snap.",
+    default: "devel" do | grade |
+    if ['stable', 'devel'].include? grade
+      grade
+    else
+      raise "Unsupported grade type '#{grade}'"
+    end
+  end
+
+   # Input a snap
+  def input(input_snap)
+    extract_snap_to_staging input_snap
+    extract_snap_metadata_from_staging
+  end # def input
+
+  # Output a snap.
+  def output(output_snap)
+    output_check(output_snap)
+
+    write_snap_yaml
+
+    # Create the snap from the staging path
+    safesystem("mksquashfs", staging_path, output_snap, "-noappend", "-comp",
+               "xz", "-no-xattrs", "-no-fragments", "-all-root")
+  end # def output
+
+  def to_s(format=nil)
+    # Default format if nil
+    # name_version_arch.snap
+    return super(format.nil? ? "NAME_FULLVERSION_ARCH.EXTENSION" : format)
+  end # def to_s
+
+  private
+
+  def extract_snap_to_staging(snap_path)
+    safesystem("unsquashfs", "-f", "-d", staging_path, snap_path)
+  end
+
+  def extract_snap_metadata_from_staging
+    metadata = YAML.safe_load(File.read(
+      staging_path(File.join("meta", "snap.yaml"))))
+
+    self.name = metadata["name"]
+    self.version = metadata["version"]
+    self.description = metadata["summary"] + "\n" + metadata["description"]
+    self.architecture = metadata["architectures"][0]
+    self.attributes[:snap_confinement] = metadata["confinement"]
+    self.attributes[:snap_grade] = metadata["grade"]
+
+    if metadata["apps"].nil?
+      attributes[:snap_apps] = []
+    else
+      attributes[:snap_apps] = metadata["apps"]
+    end
+
+    if metadata["hooks"].nil?
+      attributes[:snap_hooks] = []
+    else
+      attributes[:snap_hooks] = metadata["hooks"]
+    end
+  end
+
+  def write_snap_yaml
+    # Write the snap.yaml
+    if attributes[:snap_yaml]
+      logger.debug("Using '#{attributes[:snap_yaml]}' as the snap.yaml")
+      yaml_data = File.read(attributes[:snap_yaml])
+    else
+      summary, *remainder = (self.description or "no summary given").split("\n")
+      description = "no description given"
+      if remainder.any?
+        description = remainder.join("\n")
+      end
+
+      yaml_data = {
+        "name" => self.name,
+        "version" => self.version,
+        "summary" => summary,
+        "description" => description,
+        "architectures" => [self.architecture],
```

Comment:
If you think it is appropriate, you could convert rpm's `noarch` concept to `all` similar to how it is done in fpm for [rpm](https://github.com/jordansissel/fpm/blob/master/lib/fpm/package/rpm.rb#L231-L234) and [deb](https://github.com/jordansissel/fpm/blob/master/lib/fpm/package/deb.rb#L221-L223). This mainly to help users who float between package formats day to day. 

It is OK to to say "no" to this suggestion :)

Reply:
Ah, good idea!